### PR TITLE
feat: Email password feature branch

### PR DESCRIPTION
**Pull request by iresharma**
## Summary of change
Adds APIs and logic to the dashboard recipe to enable email password based login

## Related issues
-   https://github.com/supertokens/dashboard/issues/55

## Test Plan
- Passes all current tests
- Manual testing with dashboard repo

## Documentation changes
None

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
 
## Remaining TODOs for this PR
None

### CODE REVIEW

File: supertokens_python/constants.py [hunk 1]
```diff
@@ -11,7 +11,18 @@
...
+    "2.10",
+    "2.11",
+    "2.12",
+    "2.13",
+    "2.14",
+    "2.15",
+    "2.16",
+    "2.17",
+    "2.18",
+]
```
Comment (nkshah2) [inline supertokens_python/constants.py:25]: DASHBOARD_VERSION should be 0.4

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 2]
```diff
@@ -14,16 +14,17 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
-from typing import TYPE_CHECKING, Any, Dict, Union, Callable, Awaitable, Optional
+from typing import TYPE_CHECKING, Any, Awaitable, Callable, Dict, Optional, Union
```
Comment (nkshah2) [inline supertokens_python/recipe/emailverification/interfaces.py:17]: Reset this file

File: supertokens_python/recipe/passwordless/interfaces.py [hunk 3]
```diff
@@ -14,24 +14,27 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
-from typing import Any, Dict, List, Union
+from typing import TYPE_CHECKING, Any, Dict, List, Union
```
Comment (nkshah2) [inline supertokens_python/recipe/passwordless/interfaces.py:17]: Reset this file

File: supertokens_python/supertokens.py [hunk 4]
```diff
@@ -49,10 +49,10 @@
 from .utils import (
     execute_async,
     get_rid_from_header,
+    get_top_level_domain_for_same_site_resolution,
```
Comment (nkshah2) [inline supertokens_python/supertokens.py:52]: Reset this file

File: tests/auth-react/django3x/mysite/utils.py [hunk 5]
```diff
@@ -2,6 +2,8 @@
 from typing import Any, Dict, List, Optional, Union
 
 from dotenv import load_dotenv
+from typing_extensions import Literal
```
Comment (nkshah2) [inline tests/auth-react/django3x/mysite/utils.py:5]: Reset unrelated changes in this file

File: tests/auth-react/fastapi-server/app.py [hunk 6]
```diff
@@ -44,7 +46,7 @@
     thirdpartypasswordless,
     userroles,
 )
-from supertokens_python.recipe.dashboard import DashboardRecipe
+from supertokens_python.recipe.dashboard.recipe import DashboardRecipe
```
Comment (nkshah2) [inline tests/auth-react/fastapi-server/app.py:49]: Reset unrelated changes in this file

File: tests/auth-react/flask-server/app.py [hunk 7]
```diff
@@ -17,6 +17,8 @@
 from dotenv import load_dotenv
 from flask import Flask, g, jsonify, make_response, request
 from flask_cors import CORS
+from typing_extensions import Literal
```
Comment (nkshah2) [inline tests/auth-react/flask-server/app.py:20]: Reset unrelated changes in this file

File: tests/utils.py [hunk 8]
```diff
@@ -13,22 +13,22 @@
...
 from datetime import datetime, timezone
-from urllib.parse import unquote
 from http.cookies import SimpleCookie
 from os import environ, kill, remove, scandir
 from shutil import rmtree
 from signal import SIGTERM
 from subprocess import DEVNULL, run
 from time import sleep
 from typing import Any, Dict, List, cast
+from urllib.parse import unquote
```
Comment (nkshah2) [inline tests/utils.py:23]: Reset unrelated changes in this file

File: supertokens_python/constants.py [hunk 9]
```diff
@@ -11,7 +11,18 @@
...
+    "2.9",
+    "2.10",
+    "2.11",
+    "2.12",
+    "2.13",
+    "2.14",
+    "2.15",
+    "2.16",
+    "2.17",
+    "2.18",
```
Comment (nkshah2) [inline supertokens_python/constants.py:24]: Update the VERSION constant as well

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 2 above)
Comment (iresharma) [reply supertokens_python/recipe/emailverification/interfaces.py:17]: Changed because of circular imports

File: supertokens_python/recipe/passwordless/interfaces.py (hunk 3 above)
Comment (iresharma) [reply supertokens_python/recipe/passwordless/interfaces.py:17]: Changed because of circular imports

File: tests/auth-react/django3x/mysite/utils.py (hunk 5 above)
Comment (iresharma) [reply tests/auth-react/django3x/mysite/utils.py:5]: `dashboard/__iniy__.py` doesn't globally import `dashboard/recipe.py` so these imports change

File: tests/auth-react/fastapi-server/app.py (hunk 6 above)
Comment (iresharma) [reply tests/auth-react/fastapi-server/app.py:49]: `dashboard/__iniy__.py` doesn't globally import `dashboard/recipe.py` so these imports change

File: tests/auth-react/flask-server/app.py (hunk 7 above)
Comment (iresharma) [reply tests/auth-react/flask-server/app.py:20]: `dashboard/__iniy__.py` doesn't globally import `dashboard/recipe.py` so these imports change

File: tests/utils.py (hunk 8 above)
Comment (iresharma) [reply tests/utils.py:23]: `dashboard/__iniy__.py` doesn't globally import `dashboard/recipe.py` so these imports change

**Review by nkshah2 — APPROVED** [review]

File: supertokens_python/recipe/dashboard/__init__.py [hunk 10]
```diff
@@ -14,18 +14,20 @@
...
+    from supertokens_python.recipe.dashboard.utils import InputOverrideConfig
 
 
 def init(
-    api_key: str,
+    api_key: Union[str, None] = None,
     override: Optional[InputOverrideConfig] = None,
 ) -> Callable[[AppInfo], RecipeModule]:
+    # Global import for the following was avoided because of circular import errors
+    from .recipe import DashboardRecipe
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/dashboard/__init__.py:29]: but there is no more an issue of circular dependency right? Cause we are not doing auto init of dashboard recipe. So remove this comment and make it a regular global import?

File: supertokens_python/recipe/dashboard/recipe_implementation.py [hunk 11]
```diff
@@ -33,12 +34,21 @@ async def should_allow_access(
...
-        api_key_header_value = request.get_header("authorization")
-
-        # We receive the api key as `Bearer API_KEY`, this retrieves just the key
-        api_key = api_key_header_value.split(" ")[1] if api_key_header_value else None
-
-        if api_key is None:
-            return False
-
-        return api_key == config.api_key
+        if not config.api_key:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/dashboard/recipe_implementation.py:37]: use auth_mode

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 12]
```diff
@@ -14,16 +14,17 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:16]: undo changes in this file

File: supertokens_python/recipe/passwordless/interfaces.py [hunk 13]
```diff
@@ -14,24 +14,27 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/passwordless/interfaces.py:16]: undo changes in this file

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/dashboard/__init__.py (hunk 10 above)
Comment (iresharma) [reply supertokens_python/recipe/dashboard/__init__.py:29]: ok

**Review by rishabhpoddar — APPROVED** [review]
